### PR TITLE
nix: fixup sandbox dns resolution failure

### PR DESCRIPTION
**Pull request by baloo**
###### Motivation for this change

see https://github.com/NixOS/nixpkgs/pull/135689
https://github.com/NixOS/nix/issues/5089
https://github.com/NixOS/nix/pull/5224
https://github.com/NixOS/nix/pull/5228

Nix just merged fixes for DNS resolution failure. This should fix build errors on r13y and hydra.

This also introduces a test that check for the sandbox.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/tests/nix-fetch.nix [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        client.succeed("nix --version >&2")
+        client.succeed("curl -vvv http://example.com/index.html >&2")
+
+    with subtest("nix-build can lookup dns and fetch data"):
+        client.succeed("""
+          nix-build -E 'derivation {
+              builder = "builtin:fetchurl";
+
+              # New-style output content requirements.
+              outputHash = "0ix4jahrkll5zg01wandq78jw3ab30q4nscph67rniqg5x7r0j59";
```
Comment (Mindavi) [inline nixos/tests/nix-fetch.nix:82]: Is the hash of example.com stable? Maybe a different stable url would be better?

Comment (baloo) [reply nixos/tests/nix-fetch.nix:82]: this only run in the test environment and this is the hash of hello world:
```
$ nix-hash --type sha256 --to-base32 $(echo 'hello world' | sha256sum | cut -d ' ' -f 1)
0ix4jahrkll5zg01wandq78jw3ab30q4nscph67rniqg5x7r0j59
```

The file is created line 37.

Comment (baloo) [reply nixos/tests/nix-fetch.nix:89]: I pushed a rewritten version making that explicit.

Comment (Mindavi) [reply nixos/tests/nix-fetch.nix:89]: I read that earlier and totally forgot. Then it's indeed all good :)

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 137197` at 0dfe0346 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ovftool</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/d0f17ae369443aa6cceaeb68dbce3942">hydra-unstable</a></li>
    <li><a href="https://gist.github.com/ddc3249dee60075aadfd0de0fc227cf3">nix-update</a></li>
    <li><a href="https://gist.github.com/cf6f37d111c23306f5700a7e4784c447">nixUnstable</a></li>
    <li><a href="https://gist.github.com/11abec7a0e39f6e7f123a341945797c5">nixpkgs-review</a></li>
  </ul>
</details>
<details>
  <summary>1 package skipped due to time constraints:</summary>
  <ul>
    <li>tests.trivial</li>
  </ul>
</details>
<details>
  <summary>60 packages built successfully:</summary>
  <ul>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cachix</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haskellPackages.cachix</li>
    <li>haskellPackages.hercules-ci-agent</li>
    <li>haskellPackages.hercules-ci-cnix-expr</li>
    <li>haskellPackages.hercules-ci-cnix-store</li>
    <li>haskellPackages.nix-paths</li>
    <li>haskellPackages.nvfetcher</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>hercules-ci-agent</li>
    <li>hubstaff</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix (nixStable)</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nixos-generators</li>
    <li>nixos-install-tools</li>
    <li>nixos-option</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixui</li>
    <li>nvfetcher</li>
    <li>python38Packages.nix-kernel</li>
    <li>python38Packages.nixpkgs</li>
    <li>python38Packages.pythonix</li>
    <li>python39Packages.nix-kernel</li>
    <li>python39Packages.nixpkgs</li>
    <li>python39Packages.pythonix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>vulnix</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-uri-literals.md">no-uri-literals</a>

URI literals are deprecated.
Near pkgs/tools/package-management/nix/default.nix:211:15:
```
    |
211 |         url = https://github.com/hercules-ci/nix/raw/5c58d84a76d96f269e3ff1e72c9c9ba5f68576af/boehmgc-coroutine-sp-fallback.diff;
    |               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/tools/package-management/nix/default.nix:74:17:
```
   |
74 |                 ./aws-sdk-cpp-TransferManager-ContentEncoding.patch
   |                 ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/tools/package-management/nix/default.nix:234:7:
```
    |
234 |       (fetchpatch {
    |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/tools/package-management/nix/default.nix:258:7:
```
    |
258 |       (fetchpatch {
    |       ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by baloo** [comment]
can't reproduce those 4 build failures here.

**Comment by baloo** [comment]
I can't reproduce the build failures:
```
$ git checkout 0dfe034656fdf2e924f3163c25b60d6338fd2951
$ nix-build . --check -A nix.patches
checking outputs of '/nix/store/imvc9ngc7y68vi9ahvzlin9ds6pwi11n-5228.patch.drv'...

trying https://patch-diff.githubusercontent.com/raw/NixOS/nix/pull/5228.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6456    0  6456    0     0  18340      0 --:--:-- --:--:-- --:--:-- 18340
/nix/store/hq620q0nsmdjbn0mmzblznqzkpzxkn28-5228.patch
$ nix-build . --check -A nixUnstable.patches
checking outputs of '/nix/store/b1dp6q3r16mvs5hy7js835y4700z1pnl-5224.patch.drv'...

trying https://patch-diff.githubusercontent.com/raw/NixOS/nix/pull/5224.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6757    0  6757    0     0  14468      0 --:--:-- --:--:-- --:--:-- 14468
/nix/store/qnf0304vg8l214kw1b2lq94s707l111m-5224.patch
```

**Comment by baloo** [comment]
Result of `nixpkgs-review pr 137197` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ovftool</li>
  </ul>
</details>
<details>
  <summary>3 packages blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
    <li>tests.nixos-functions.nixosTest-test</li>
    <li>tests.trivial</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>63 packages built:</summary>
  <ul>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cachix</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haskellPackages.cachix</li>
    <li>haskellPackages.hercules-ci-agent</li>
    <li>haskellPackages.hercules-ci-cnix-expr</li>
    <li>haskellPackages.hercules-ci-cnix-store</li>
    <li>haskellPackages.nix-paths</li>
    <li>haskellPackages.nvfetcher</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>hercules-ci-agent</li>
    <li>hubstaff</li>
    <li>hydra-unstable</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix (nixStable)</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nixUnstable</li>
    <li>nixos-generators</li>
    <li>nixos-option</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>nixui</li>
    <li>nvfetcher</li>
    <li>python38Packages.nix-kernel</li>
    <li>python38Packages.nixpkgs</li>
    <li>python38Packages.pythonix</li>
    <li>python39Packages.nix-kernel</li>
    <li>python39Packages.nixpkgs</li>
    <li>python39Packages.pythonix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>vulnix</li>
  </ul>
</details>

**Comment by baloo** [comment]
not entirely sure why there are inconsistencies in the nixpkgs-reviews here :/

**Comment by L-as** [comment]
Does any issue depend on this being merged?

**Comment by baloo** [comment]
https://www.r13y.com has been showing broken builds for a month now.
Sounds like hydra builders have regular issue with it too: https://github.com/NixOS/nix/issues/5089#issuecomment-917606674

**Review by L-as — APPROVED** [review]
LGTM

**Comment by edolstra** [comment]
Rather than apply a patch, isn't it better to just bump nixUnstable to the latest version?

**Comment by baloo** [comment]
I'll change that, but that does not fix nix2.3

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 137197` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ovftool</li>
  </ul>
</details>
<details>
  <summary>3 packages blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
    <li>tests.nixos-functions.nixosTest-test</li>
    <li>tests.trivial</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>64 packages built:</summary>
  <ul>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cachix</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dep2nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haskellPackages.cachix</li>
    <li>haskellPackages.hercules-ci-agent</li>
    <li>haskellPackages.hercules-ci-cnix-expr</li>
    <li>haskellPackages.hercules-ci-cnix-store</li>
    <li>haskellPackages.nix-paths</li>
    <li>haskellPackages.nvfetcher</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>hercules-ci-agent</li>
    <li>hubstaff</li>
    <li>hydra-unstable</li>
    <li>libnixxml</li>
    <li>lispPackages.quicklisp-to-nix</li>
    <li>lispPackages.quicklisp-to-nix-system-info</li>
    <li>nix (nixStable)</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-direnv-flakes</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-simple-deploy</li>
    <li>nix-template</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixStatic</li>
    <li>nixUnstable</li>
    <li>nixos-generators</li>
    <li>nixos-option</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>nixui</li>
    <li>nvfetcher</li>
    <li>python38Packages.nix-kernel</li>
    <li>python38Packages.nixpkgs</li>
    <li>python38Packages.pythonix</li>
    <li>python39Packages.nix-kernel</li>
    <li>python39Packages.nixpkgs</li>
    <li>python39Packages.pythonix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>vulnix</li>
  </ul>
</details>

File: pkgs/tools/package-management/nix/default.nix [hunk 2]
```diff
@@ -229,18 +229,26 @@ in rec {
...
+        sha256 = "sha256-irK1XCsXFwJA4tYzycaGIttiiJdgwbldSBFLBR/ImTk=";
+      })
+    ];
   });
 
   nixUnstable = lib.lowPrio (callPackage common rec {
     pname = "nix";
     version = "2.4${suffix}";
-    suffix = "pre20210908_${lib.substring 0 7 src.rev}";
+    suffix = "pre20210915_d2c8eed";
```
Comment (Mindavi) [inline pkgs/tools/package-management/nix/default.nix:245]: Why stop using the substring function?

Comment (baloo) [reply pkgs/tools/package-management/nix/default.nix:245]: a bad rebase / oversight.

Comment (baloo) [reply pkgs/tools/package-management/nix/default.nix]: Thanks for catching that, I fixed

**Comment by wizeman** [comment]
This has solved a few issues I was having with the Nix daemon on the `release-21.05` branch.

First, I was running into NixOS/nix#5089, of course.

I applied the `nix.sandboxPaths = [ "/var/run/nscd/socket" ];` workaround, but then I started running into the issue described in the following commit on my remote builders: https://github.com/NixOS/nix/commit/11882d7c7ce3b6dc51dd7c0536f9662dc254ac0a .

This happened when running the `installCheck` phase of building the `nixStable` package. However, that commit (which supposedly fixes the `No user exists for uid ...` error) seems to only be applied on `nixUnstable`, but not `nixStable`.

I then tried running `nixUnstable` on my remote builders, but then I started running into NixOs/nix#5270.

(Backporting) this PR has fixed all of the above issues for me, as far as I can tell.

**Review by r-burns — APPROVED** [review]
Test fails on the commit introducing it, and passes with the patched nix.

**Comment by baloo** [comment]
nixUnstable was bumped in https://github.com/NixOS/nixpkgs/pull/138623 which created a merge conflict.

**Comment by tomberek** [comment]
I've lost track of the current status of the situation. Is this still needed? What versions work, what doesn't?

**Comment by baloo** [comment]
nixUnstable have been bumped, and is now fixed. We still need nixStable to be fixed. Either merge this, or get a new nix2.3 release and deploy it.

This still brings a regression test which I think is useful.

**Review by trofi — APPROVED** [review]

**Pull request closed without merging**